specs/agents/tracing-instrumentation-aws-lambda.md

```diff
@@ -133,15 +133,43 @@ In version 2.0, the `${event.requestContext.routeKey}` can have the format `GET
 
-With only one message in `event.records`, the agents can use the single SQS / SNS `record` to retrieve the `traceparent` and `tracestate` from `record.messageAttributes` and use it for starting the lambda transaction.
+SQS or SNS events that include **exactly one** message are treated as special case: more context is reported, and trace-context (if present) is handled differently.
 
-In addition the following fields should be set for Lambda functions triggered by SQS or SNS:
+#### SQS event with multiple messages
+
+TODO: My read of https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html is that a given trigger invocation can only include messages from one queue. Therefore we can use the eventSourceArn from the *first* record for the transaction name.  Also, inferring from https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#batch-receiving OTel implies that a transaction name including the queue name would be nice.
+
+Agents SHOULD check each record for a `traceparent` message attribute, and
+create a [span link](span-links.md) on the transaction for each message with
+trace-context.
+
+In addition to [the generic Lambda transaction fields](#generic-lambda-instrumentation)
+the following fields should be set:
+
+Field | Value | Description | Source
+---   | ---   | ---         | ---
+`type` | `messaging`| Transaction type: constant value for SQS. | -
+`name` | e.g. `RECEIVE SomeQueue` | Transaction name: Follow the [messaging spec](./tracing-instrumentation-messaging.md) for transaction naming. | Simple queue name can be derived from the 6th segment of `records[0].eventSourceArn`.
+`faas.trigger.type` | `pubsub` | Constant value for message based triggers | -
+`context.service.origin.name` | e.g. `my-queue` | SQS queue name | Simple queue name can be derived from the 6th segment of `records[0].eventSourceArn`.
+`context.service.origin.id` | e.g. `arn:aws:sqs:us-east-2:123456789012:my-queue` | SQS queue ARN. | `records[0].eventSourceArn`
+`context.cloud.origin.service.name` | `sqs` | Fix value for SQS. | -
+`context.cloud.origin.region` | e.g. `us-east-1` | SQS queue region. | `records[0].awsRegion`
+`context.cloud.origin.account.id` | e.g. `12345678912` | Account ID of the SQS queue. | Parse account segment (5th) from `records[0].eventSourceArn`.
+`context.cloud.origin.provider` | `aws` | Use `aws` as fix value. | -
+
+### SQS event with a single message
+
+Agents SHOULD check the single message's message attributes for `traceparent`
+and `tracestate`. If present, these should be propagated on the transaction
+to continue a distributed trace.
+
+In addition to [the generic Lambda transaction fields](#generic-lambda-instrumentation)
+the following fields should be set:
```

Comment:
I think we should treat it the same as "SQS event with multiple messages". It seems weird that tracing behaves differently depending on how many items are in the batch.

Reply:
Thanks. We'd also discussed at https://elastic.slack.com/archives/C9A6ANR5Y/p1647477818192449 last week and again today on the apm-serverless call and agreed to drop the special-casing of a "single message in a batch" for Lambda instrumentation.

commit 07719d0